azkaban-common/src/main/java/azkaban/executor/ExecutorManager.java

```diff
@@ -371,18 +672,34 @@ public JobMetaData getExecutionJobMetaData(ExecutableFlow exFlow,
     }
   }
 
+  /**
+   * if flows was dispatched to an executor, cancel by calling Executor else if
+   * flow is still in queue, remove from queue and finalize {@inheritDoc}
+   *
+   * @see azkaban.executor.ExecutorManagerAdapter#cancelFlow(azkaban.executor.ExecutableFlow,
+   *      java.lang.String)
+   */
   @Override
   public void cancelFlow(ExecutableFlow exFlow, String userId)
-      throws ExecutorManagerException {
+    throws ExecutorManagerException {
     synchronized (exFlow) {
-      Pair<ExecutionReference, ExecutableFlow> pair =
+      if (runningFlows.containsKey(exFlow.getExecutionId())) {
+        Pair<ExecutionReference, ExecutableFlow> pair =
           runningFlows.get(exFlow.getExecutionId());
-      if (pair == null) {
+        callExecutorServer(pair.getFirst(), ConnectorParams.CANCEL_ACTION,
+          userId);
+      } else if (queuedFlowMap.containsKey(exFlow.getExecutionId())) {
```

Comment:
just thinking, do you think we should check in the queuedFlowMap first, then check in the runningFlows?

Otherwise there is the very very very slim chance of when the check is being performed on RunningFlows, it was being taken out of the queue, and when we check the queue, it is already in the RunningFlows